fms_extras/utils/generation.py

```diff
@@ -3,9 +3,532 @@
+            for head n. len(threshes) must equal speculator.n_predict. prod(threshes) must be greater
+            than or equal to n_candidates (we cannot have more candidates than tree leaves).
+        flatting: enable batch flattening / tree attention with redundant prefix removal when compression
+            ratio is favorable. Adds extra overhead to shrink the token count in each batch.
+        decode_model: nn.Module, optional
+            an optional model that performs the decode forward step. If None, will use the function or nn.Module
+            provided to model param to perform decode forward. (default is None)
+        cudagraphs: bool
+            if True, cudagraphs is used and all metadata will be padded, otherwise metadata will not be padded unless
+            required. Note: This is a WIP and only works for batch_size=1
+    Returns:
+        result: List of id tensors, possibly different lengths if batching.
+        n_steps: Number of foward passes used to generate provided tokens.
+    """
+    start_time = time.time()
+    # Construct batch(es) and initial inputs
+    if decode_model is None:
+        decode_model = model
+    bsize = len(input_ids)
+
+    if cudagraphs and (flatting or bsize != 1):
+        raise NotImplementedError(
+            "cudagraphs is not yet supported for batch sizes greater than 1 or flatting"
+        )
+
+    result = input_ids  # [b] n
+    inp_len = speculator.n_predict + 1
+
+    # get the last tokens if past max length in model
+    input_ids = [seq[-max_seq_len + inp_len :] for seq in input_ids]
+    model_input_lengths = [seq.size(0) for seq in input_ids]
+    max_len = max(model_input_lengths)
+
+    # prepare model input for prefill
+    inputs, cache_data, mask = __prepare_inputs_for_prefill(
+        input_ids,
+        kv_cache_manager,
+        model_input_lengths,
+        max_len,
+    )
+
+    # perform prefill
+    output = model(
+        inputs,
+        mask=mask,
+        cache_data=cache_data,
+        return_embeds=True,
+        use_cache=True,
+    )
+    logits, _, embeds = output
+    embeds = embeds[:, -1:]  # b 1 d
+    logits = logits[:, -1:]  # b 1 v
+
+    # get the time to first token
+    ttft = time.time() - start_time
+
+    start_time = time.time()
+
+    parent_sequence_ids = cache_data.sequence_ids
+    n_gen = torch.zeros(bsize, device=inputs.device, dtype=torch.int)
+    n_steps = 0
+    input_ids = torch.argmax(logits, dim=2)  # b 1
+    result = [
+        torch.cat((line, input_id), dim=0)
+        for line, input_id in zip(result, list(input_ids))
+    ]
+    block_mapping_max = ((max_len + new_tokens + inp_len - 1) // 16) + 1
+    # perform decode step
+    while min(n_gen) < new_tokens:
+        n_steps += 1
+
+        model_input_lengths = [inp_len for _ in range(input_ids.size(0) * n_candidates)]
+
+        # allocate candidate sequences in cache and get the sequence ids
+        (
```

Comment:
nit: this formatting, leading with an open-paren, is a bit hard to read. might want to consider splitting it up some how. E.g. maybe something like this might make it easier to see what's going on?
```
result = function_call(blahblah)
a, b, c = result
```
then again, is parent_sequence_ids even used? I see it reset below by the call to __free_incorrect_tokens_and_sequences

likewise it seems `model_input_lengths` is set above, overriden here, then never used? that makes me further wonder if we've correctly refactored out the right bits of code to make this function more readable. we should avoid returning tuples like this if not necessary, since they can be confusing to read

Reply:
Yes, `parent_sequence_ids` need to be updated on each step to then be used with:

```
# allocate candidate sequences in cache and get the sequence ids
        (
            cache_data,
            child_sequence_ids_list,
            child_sequence_ids_flattened,
        ) = __allocate_candidate_sequences(
            kv_cache_manager, parent_sequence_ids, model_input_lengths, n_candidates
        )
```

Similarly, `model_input_lengths` is used with the same call to allocate

Reply:
err... maybe I just copy/pasted the wrong variable name into my comment. I don't see `child_sequence_ids_flattened` used anywhere?

Reply:
Yes, you are right, removing this from the return

Reply:
@nairbv Fixed in latest commit